README.md
# hydrusrv [![Build status][travis-badge]][travis] [![Docker Hub build][docker-hub-badge]][docker-hub] [![Known vulnerabilities][snyk-badge]][snyk] [![JavaScript Standard Style][standardjs-badge]][standardjs]

> A small application for serving media managed with
> [hydrus server][hydrus-server] via API

__Adandoned: hydrusrv is no longer maintained.__

hydrusrv is a small application based on [Express][express] that can serve
media managed with [hydrus server][hydrus-server] over a REST-like API. A
separate application called [hydrusrv-sync][hydrusrv-sync] is used to access
the databases of hydrus server and copy over the data into hydrusrv's content
database.

hydrusrv provides its own (simple) user handling and token-based
authentication (stored in a separate database) instead of relying on hydrus
server's access keys.

A [Vue][vue]-based Web client for hydrusrv called [hydrusrvue][hydrusrvue] is
also available.

## Table of contents

+ [Install](#install)
  + [Dependencies](#dependencies)
  + [Updating](#updating)
    + [Upgrading from 7.x to 8.x](#upgrading-from-7x-to-8x)
    + [Upgrading from 6.x to 7.x](#upgrading-from-6x-to-7x)
    + [Upgrading from 5.x to 6.x](#upgrading-from-5x-to-6x)
    + [Upgrading from 4.x to 5.x](#upgrading-from-4x-to-5x)
    + [Upgrading from 3.x to 4.x](#upgrading-from-3x-to-4x)
    + [Upgrading from 2.x to 3.x](#upgrading-from-2x-to-3x)
    + [Upgrading from 1.x to 2.x](#upgrading-from-1x-to-2x)
+ [Usage](#usage)
  + [Configuration](#configuration)
  + [Running the server](#running-the-server)
    + [Running with Docker](#running-with-docker)
+ [API](#api)
+ [Attribution](#attribution)
+ [Donate](#donate)
+ [Maintainer](#maintainer)
+ [Contribute](#contribute)
+ [License](#license)

## Install

The easiest way to install is via cloning this repository:

```zsh
user@local:~$ git clone https://github.com/mserajnik/hydrusrv.git
user@local:~$ cd hydrusrv
user@local:hydrusrv$ yarn
```

If you encounter any errors during installation, those are likely caused by
packages hydrusrv uses that utilize native bindings (e.g.,
[better-sqlite3][better-sqlite3] and [phc-argon2][phc-argon2]).

Please check the error message and contact the maintainers of those packages
directly if you cannnot resolve your issues.

### Dependencies

+ [hydrusrv-sync][hydrusrv-sync] (`3.x` for hydrusrv `8.x`)
+ [Node.js][node-js]
+ [Yarn][yarn]

I usually use the latest versions of Node.js and Yarn; if there has not been an
update to this repository in a while and something breaks on the latest
Node/Yarn version, please [let me know][issues].

### Updating

If you have installed via cloning the repository, you can update via Git:

```zsh
user@local:hydrusrv$ git pull
user@local:hydrusrv$ yarn
user@local:hydrusrv$ yarn migrate
```

Always make sure to run `yarn` after updating to install any packages you might
be missing. `yarn migrate` updates your authentication database with the latest 
changes.

hydrusrv follows [semantic versioning][semantic-versioning] and any breaking
changes that require additional attention will be released under a new major
version (e.g., `2.0.0`). Minor version updates (e.g., `1.1.0` or `1.2.0`) are
therefore always safe to simply install via the routine mentioned before.

When necessary, this section will be expanded with upgrade guides to new major
versions.

#### Upgrading from 7.x to 8.x

Upgrading from `7.x` to `8.x` can be done via
`git pull && yarn && yarn migrate` and requires no further manual changes.

If you are using hydrusrvue, you need to upgrade it to `5.x` to maintain
compatibility with hydrusrv `8.x`.

#### Upgrading from 6.x to 7.x

Upgrading from `6.x` to `7.x` can be done via
`git pull && yarn && yarn migrate` and requires no further manual changes.

If you are using hydrusrvue, you need to upgrade it to `4.x` to maintain
compatibility with hydrusrv `7.x`.

#### Upgrading from 5.x to 6.x

Upgrading from `5.x` to `6.x` can be done via `git pull && yarn` and requires
no further manual changes.

If you are using hydrusrv-sync and hydrusrvue, you need to upgrade those to
`3.x` to maintain compatibility with hydrusrv `6.x`.

#### Upgrading from 4.x to 5.x

Upgrading from `4.x` to `5.x` can be done via `git pull && yarn` and requires
manual updates/changes in `.env`.

If you are using hydrusrv-sync and hydrusrvue, you need to upgrade those to
`2.x` to maintain compatibility with hydrusrv `5.x`.

#### Upgrading from 3.x to 4.x

Upgrading from `3.x` to `4.x` can be done via `git pull && yarn` and requires
manual updates/changes in `.env`. You also need to make additional manual
adjustments as described in the following upgrade guide:

Starting with `4.0.0`, hydrusrv no longer copies the hydrus server data over by
itself, but relies on a separate application called
[hydrusrv-sync][hydrusrv-sync] to do so.

This change was mainly made to increase performance. Before, hydrusrv would
hold all the copied data in-memory, only allowing for a single thread to access
it. Due to the SQLite library that's being used only featuring a synchronous
API, a single request or a sync task could block the whole API.

With `4.x`, hydrusrv-sync persists the data and hydrusrv accesses it using
multiple workers (by default, one for every logical CPU core, but that is
configurable), multiplying the possible requests per seconds since a single
requests will only block a single worker for a very short period.

In addition, hydrusrv now uses two databases – the authentication database and
the content database. While the authentication database is basically what the
old app database was and is still managed completely by hydrusrv itself using
migrations, the content database holds the hydrus server data and is managed by
hydrusrv-sync, hydrusrv itself only reads from it.

Therefore, you have to install hydrusrv-sync and either run it manually every
time you want to sync the data from hydrus server or set up a cron job
(recommended).

Renaming the old `app.db` to `authentication.db` should work so you do not have
to migrate the users manually.

#### Upgrading from 2.x to 3.x

Upgrading from `2.x` to `3.x` can be done via `git pull && yarn`. You also
need to make additional manual adjustments as described in the following
upgrade guide:

The app database structure has changed, with the table `tokens` having a new
column `media_hash`. You need to either re-create your database from the
template or add this column to your existing database manually.

It is __strongly recommended__ to re-create the database by copying the
template and running migrations (`yarn migrate`). This will ensure that any
future changes will just require running migrations instead of making manual
adjustments or starting over with an empty database.

#### Upgrading from 1.x to 2.x

Upgrading from `1.x` to `2.x` can be done via `git pull && yarn` and requires
manual updates/changes in `.env`.

## Usage

### Configuration

After installing, the first thing you want to do is duplicating both the
authentication database template and the content database template you can find
under `storage/authentication.db.template` and `storage/content.db.template`.
These [SQLite][sqlite] databases are used to store both the users and tokens
for authentication, as well as a copy of the hydrus server data in an optimized
form (for faster on-demand querying).

The default (and recommended) location of the databases is
`storage/authentication.db` and `storage/content.db`, but you are free to put
them wherever you want and can rename them to your liking.

Next, you also need to duplicate `.env.example` to `.env`. This file is used to
configure hydrusrv and needs to be located in the root directory of the
application.

After copying it, you can edit `.env` and change hydrusrv's configuration. The
following are all the available settings (along with their default values):

+ `NODE_ENV=development`: defines the environment hydrusrv is running in.
  It currently does not affect anything besides the access logging but it
  should be set to `production` in a live environment and `development` when
  developing.
+ `URL=https://example.com`: the URL under which hydrusrv is accessible. Used
  to generate media URLs. __No trailing slashes.__
+ `PORT=8000`: the port hydrusrv is listening on. This can be different than
  the port used to access it from outside when proxying via [nginx][nginx]
  (recommended) or similar solutions.
+ `API_BASE=/api`: the base path of all the API routes. __No trailing__
  __slashes.__
+ `MEDIA_BASE=/media`: the base path of all the media routes. __No trailing__
  __slashes.__
+ `CROSS_ORIGIN_ALLOWED=false`: allows cross-origin requests (useful when the
  application accessing the API is located on a different domain).
+ `AUTHENTICATION_DB_PATH=./storage/authentication.db`: the authentication
  database path (absolute or relative). The database must exist and the file
  must be read-/writeable for hydrusrv.
+ `CONTENT_DB_PATH=./storage/content.db`: the content database path (absolute
  or relative). The database must exist and the file must be read-/writeable
  for hydrusrv.
+ `HYDRUS_FILES_PATH=`: sets the path to the hydrus server files directory
  (called `server_files`).  __Absolute path required.__
+ `NUMBER_OF_WORKERS=`: sets the number of workers. By default, one worker per
  logical CPU core is used. You might want to decrease or increase that number,
  depending on your needs/hardware. In general, the more workers are running,
  the more requests can be handled simultaneously. But note that increasing the
  number of workers beyond the number of logical CPUs might be detrimental to
  performance or cause even more serious issues (e.g., crashes).
+ `DB_CHECKPOINT_INTERVAL=3600`: sets the interval (in seconds) at which
  hydrusrv [checkpoints][checkpoint] the authentication database (the content
  database is handled by hydrusrv-sync).
+ `REGISTRATION_ENABLED=true`: setting this to `false` disables the creation of
  new users.
+ `AUTHENTICATION_REQUIRED=true`: setting this to `false` allows the access of
  all non-authentication-related routes without providing a (media) token,
  effectively making the API open for anyone to access. This does not disable
  authentication-related routes altogether, it merely makes authentication
  optional.
+ `MIN_PASSWORD_LENGTH=16`: sets the minimum password length when creating or
  updating users. Providing a higher value than `1024` will discard the value
  and use `1024` as the minimum length instead.
+ `FILES_PER_PAGE=42`: the results per page when listing files.
+ `TAGS_PER_PAGE=42`: the results per page when listing tags.
+ `AUTOCOMPLETE_LIMIT=10`: the maximum amount of tag completion results.
+ `COUNTS_ENABLED=false`: enables the output of total counts when listing files
  and tags for the cost of response times (especially with larger databases).
+ `COUNTS_CACHING_ENABLED=false`: setting this to `true` enables the caching of
  file/tag counts (only relevant when `COUNTS_ENABLED` is `true`). This is
  recommended for larger databases to decrease response times when queries are
  made that only differ in page and sorting since the count will only need to
  be calculated once (for the first query). On smaller databases, the
  performance gain might not be noticeable (or it might even be slighty slower
  for very fast queries). The cache is emptied when hydrusrv-sync runs.
+ `ACCESS_LOGGING_ENABLED=false`: setting this to `false` disables access
  logging when
  `NODE_ENV=production` is set.
+ `ACCESS_LOGFILE_PATH_OVERRIDE=`: overrides the default access logfile
  location (`logs/access.log`. Logging to a file is only enabled with
  `ACCESS_LOGGING_ENABLED=true` and `NODE_ENV=production`. With
  `NODE_ENV=development`, hydrusrv logs to the console instead.
  __Absolute path required.__

After you are done making adjustments, make sure `AUTHENTICATION_DB_PATH`
points to the correct location and run migrations to finish setting up the
authentication database:

```zsh
user@local:hydrusrv$ yarn migrate
```

The content database does not have any migrations since it is entirely managed
by hydrusrv-sync.

### Running the server

First of all, you have several options to start the server:

+ `yarn start-dev`: starts the server in development mode using
  [nodemon][nodemon] to watch for file changes and restarts when necessary.
+ `yarn start`: starts the server in production mode.
+ `./bin/www`: the script that is run when using both, `start-dev` and `start`.
  You can also execute it directly to start the server in production mode.

For running in production mode, you will likely want to set up both a reverse
proxy (I recommend [nginx][nginx]) and a way to autostart hydrusrv when booting
your machine (I personally use [Supervisor][supervisor]).

### Running with Docker

You can also run hydrusrv with [Docker][docker] if you want to. A
[Docker image][docker-hub] is available on Docker Hub.

See [here][hydrusrv-docker] for a [Docker Compose][docker-compose] setup that
combines hydrus server, hydrusrv and hydrusrv-sync into an easy-to-use package.

If you want to create your own setup, please take a look at the
[Dockerfile](Dockerfile) and the
[entrypoint file](.docker/docker-entrypoint.sh) to figure out how to configure
it.

## API

You can find the API documentation [here](API.md).

## Attribution

Special thanks to the user [FredericaBernkastel][frederica-bernkastel] for
providing valuable input and writing some amazing performance improvements
that makes hydrusrv far more usable, especially with larger hydrus server
databases.

A big thank you also to the [developer of hydrus network][hydrus-developer] for
putting up with my questions and encouraging me to work on this project.

## Donate

If you like hydrusrv and want to buy me a coffee, feel free to donate via
PayPal:

[![Donate via PayPal][paypal-image]][paypal]

Alternatively, you can also send me BTC:

![Donate BTC][btc-image]  
`13jRyroNn8QF4mbGZxKS6mR3PsxjYTsGsu`

Donations are unnecessary, but very much appreciated. :)

## Maintainer

[mserajnik][maintainer]

## Contribute

You are welcome to help out!

[Open an issue][issues] or submit a pull request.

## License

[MIT](LICENSE.md) © Michael Serajnik

[travis]: https://travis-ci.org/mserajnik/hydrusrv
[travis-badge]: https://img.shields.io/travis/mserajnik/hydrusrv/master.svg

[docker-hub-badge]: https://img.shields.io/docker/automated/mserajnik/hydrusrv.svg

[snyk]: https://snyk.io/test/github/mserajnik/hydrusrv
[snyk-badge]: https://snyk.io/test/github/mserajnik/hydrusrv/badge.svg

[standardjs]: https://standardjs.com
[standardjs-badge]: https://img.shields.io/badge/code_style-standard-brightgreen.svg

[express]: https://expressjs.com/
[hydrus-server]: http://hydrusnetwork.github.io/hydrus
[hydrusrv-sync]: https://github.com/mserajnik/hydrusrv-sync
[vue]: https://vuejs.org/
[hydrusrvue]: https://github.com/mserajnik/hydrusrvue
[better-sqlite3]: https://github.com/JoshuaWise/better-sqlite3/wiki/Troubleshooting-installation
[phc-argon2]: https://github.com/simonepri/phc-argon2
[node-js]: https://nodejs.org/en/
[yarn]: https://yarnpkg.com/
[semantic-versioning]: https://semver.org/
[sqlite]: https://www.sqlite.org/
[nginx]: https://nginx.org/
[checkpoint]: https://www.sqlite.org/c3ref/wal_checkpoint.html
[supported-mime-types]: https://github.com/mserajnik/hydrusrv/blob/master/server/config/hydrus.js#L2-L14
[nodemon]: https://github.com/remy/nodemon
[supervisor]: http://supervisord.org/
[docker]: https://www.docker.com/
[docker-hub]: https://hub.docker.com/r/mserajnik/hydrusrv/
[hydrusrv-docker]: https://github.com/mserajnik/hydrusrv-docker
[docker-compose]: https://docs.docker.com/compose/
[vultr-cloud-compute]: https://www.vultr.com/pricing/
[hydrusrvue-screenshots]: https://github.com/mserajnik/hydrusrvue#screenshots
[frederica-bernkastel]: https://github.com/FredericaBernkastel
[hydrus-developer]: https://github.com/hydrusnetwork

[paypal]: https://www.paypal.me/mserajnik
[paypal-image]: https://www.paypalobjects.com/webstatic/en_US/i/btn/png/blue-rect-paypal-26px.png
[btc-image]: https://mserajnik.at/external/btc.png

[maintainer]: https://github.com/mserajnik
[issues]: https://github.com/mserajnik/hydrusrv/issues/new


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [8.2.0] - 2018-03-11

### Changed

+ Updated dependencies

### Removed

+ Removed FOSSA badge
+ Removed links to demo

## [8.1.0] - 2018-02-15

### Changed

+ Updated dependencies
+ Excluded `.git` directory from Docker image

## [8.0.0] - 2018-01-29

### Added

+ Added support for searching files with wildcarded tags

### Changed

+ Updated dependencies

## [7.0.0] - 2018-01-15

### Added

+ Added a route for getting users
+ Added a route for getting tokens

### Changed

+ Added various fields to the `users` and `tokens` tables and adjusted the API
  responses accordingly
+ Normalized the API responses for creating and updating users
+ Moved API documentation to separate file
+ Refactored code
+ Updated dependencies

### Fixed

+ Added the missing `hash` property to the file routes API documentation

## [6.0.0] - 2018-12-28

### Added

+ Added the ability to filter files by hydrus-like _system constraints_
+ Added the ability to sort files by `tag_count`
+ Added the `hash` field to the output of file routes
+ Added a route (`mime-types`) for fetching all available MIME types

### Changed

+ Changed checkpointing so the WAL file gets truncated
+ Updated dependencies

### Fixed

+ Fixed test `content.db` not being Git-ignored
+ Fixed test files not being cleaned up correctly (especially on error)

## [5.0.0] - 2018-12-09

### Added

+ Added ability to (optionally) output counts when listing files and tags
+ Added counts caching
+ Added basic error prevention when two requests trying to insert the same data
  happen almost exactly at the same time

### Changed

+ Normalized API responses to make expanding them easier in the future
+ Updated dependencies

## [4.4.0] - 2018-12-05

### Changed

+ Cleaned up environment variables
+ Updated dependencies

### Fixed

+ Removed unneeded module import

## [4.3.0] - 2018-12-05

### Changed

+ Improved "exclude tags" query performance
+ Improved checkpoint handling
+ Updated dependencies

### Fixed

+ Fixed issue where querying a file with no tags would responds with a
  malformed `tags` array causing hydrusrvue to crash

## [4.2.1] - 2018-12-04

### Fixed

+ Fixed regression introduced when increasing query performance in
  [`4.2.0`][4.2.0] that caused tags being fetched incorrectly

## [4.2.0] - 2018-12-03

### Changed

+ Improved performance of various queries
+ Improved async code
+ Refactored code
+ Dropped `hydrusrv` prefix from content database tables to be consistent with
  the application database naming scheme
+ Updated dependencies

## [4.1.0] - 2018-12-02

### Changed

+ Reduced Docker image size

### Fixed

+ Removed unintended `console.log()` statement
+ Fixed letter case in readme

## [4.0.0] - 2018-12-01

### Added

+ Added clustering to improve performance when multiple concurrent requests
  happen

### Changed

+ Split up application database into authentication database and content
  database (holding the users/tokens and hydrus server content respectively)
+ Switched to WAL mode and added checkpoint handling
+ Moved data update handling into a separate tool called
  [hydrusrv-sync](https://github.com/mserajnik/hydrusrv-sync)
+ Changed Docker configuration to make it configurable when starting a
  container and enabled automatic builds on Docker Hub
+ Improved application status logging
+ Updated dependencies

## [3.17.0] - 2018-11-25

### Added

+ Added an `AUTHENTICATION_REQUIRED` setting that can be used to make
  authentication optional instead of needed for accessing
  non-authentication-related routes

### Changed

+ Refactored and cleaned up code
+ Updated dependencies

## [3.16.1] - 2018-11-21

### Fixed

+ Fixed typo in data handler SQL query

## [3.16.0] - 2018-11-18

### Added

+ Made autocompletion result amount configurable

### Changed

+ Changed autocomplete sorting (exact match at the top, rest sorted by
  `file_count` descending)
+ Changed namespace sorting to account for numbers (e.g., `page:1`, `page:2`
  and `page:11` would be sorted as `1-11-2` before and `1-2-11` now)

### Fixed

+ Escaped underscores in `LIKE` queries

## [3.15.0] - 2018-11-14

### Added

+ Added profiling for temporary data updates
+ Added "Credits" section to readme

### Changed

+ Majorly increased temporary data update performance
+ Majorly increased files query performance
+ Pinned Node version in Dockerfile to prevent possible issues in the future
+ Replaced `console.log()` statements with `console.info()`

### Fixed

+ Fixed "Get files by tags sorted ascending" test
+ Fixed nodemon failing on certain Linux systems due to it not excluding
  the `node_modules` directory by default
+ Fixed license formatting

## [3.14.0] - 2018-11-10

### Changed

+ Improved "exclude tags" query performance
+ Updated dependencies

## [3.13.0] - 2018-11-08

### Changed

+ Updated dependencies

### Fixed

+ Fixed tags sometimes being assigned to the wrong files
+ Fixed namespaces not being limited to alphanumeric characters and underscores

## [3.12.1] - 2018-11-05

### Fixed

+ Fixed notice about native bindings in readme to reflect the current usage

## [3.12.0] - 2018-11-05

### Changed

+ Changed migration runner to no longer require node-sqlite
+ Cleaned up code
+ Updated dependencies

## [3.11.0] - 2018-10-24

### Changed

+ Increased performance when creating temporary data tables
+ Updated dependencies and clarified them in readme

## [3.10.0] - 2018-10-01

### Changed

+ Moved version declaration into config
+ Updated dependencies

## [3.9.0] - 2018-09-17

### Added

+ Added the ability to exclude files by prefixing the undesired tag(s) with `-`

### Changed

+ Refactored code
+ Updated dependencies

## [3.8.0] - 2018-08-30

### Added

+ Added FOSSA badge

### Changed


server/config/index.js
const path = require('path')

let authenticationDbPath = process.env.AUTHENTICATION_DB_PATH

if (authenticationDbPath.startsWith('.')) {
  authenticationDbPath = path.resolve(__dirname, '../..', authenticationDbPath)
}

let contentDbPath = process.env.CONTENT_DB_PATH

if (contentDbPath.startsWith('.')) {
  contentDbPath = path.resolve(__dirname, '../..', contentDbPath)
}

module.exports = {
  version: '8.2.0',
  url: process.env.URL,
  port: process.env.PORT || 8000,
  apiBase: process.env.API_BASE || '/api',
  mediaBase: process.env.MEDIA_BASE || '/media',
  crossOriginIsAllowed: (process.env.CROSS_ORIGIN_ALLOWED === 'true'),
  authenticationDbPath: authenticationDbPath,
  contentDbPath: contentDbPath,
  hydrusFilesPath: process.env.HYDRUS_FILES_PATH,
  numberOfWorkers: process.env.NUMBER_OF_WORKERS || require('os').cpus().length,
  dbCheckpointInterval: process.env.DB_CHECKPOINT_INTERVAL || 3600,
  registrationIsEnabled: (process.env.REGISTRATION_ENABLED === 'true'),
  authenticationIsRequired: process.env.AUTHENTICATION_REQUIRED
    ? (process.env.AUTHENTICATION_REQUIRED === 'true')
    : true,
  minPasswordLength: process.env.MIN_PASSWORD_LENGTH || 16,
  filesPerPage: process.env.FILES_PER_PAGE || 42,
  tagsPerPage: process.env.TAGS_PER_PAGE || 42,
  autocompleteLimit: process.env.AUTOCOMPLETE_LIMIT || 10,
  countsAreEnabled: (process.env.COUNTS_ENABLED === 'true'),
  countsCachingIsEnabled: (process.env.COUNTS_CACHING_ENABLED === 'true'),
  accessLoggingIsEnabled: (process.env.ACCESS_LOGGING_ENABLED === 'true'),
  accessLogfilePath: process.env.ACCESS_LOGFILE_PATH_OVERRIDE ||
    path.resolve(__dirname, '../../logs/access.log'),
  availableMimeTypes: {
    1: 'image/jpeg',
    2: 'image/png',
    3: 'image/gif',
    4: 'image/bmp',
    9: 'video/x-flv',
    14: 'video/mp4',
    18: 'video/x-ms-wmv',
    20: 'video/x-matroska',
    21: 'video/webm',
    23: 'image/apng',
    25: 'video/mpeg',
    26: 'video/quicktime',
    27: 'video/x-msvideo'
  }
}
